src/mgr/ActivePyModules.cc

```diff
@@ -382,27 +382,30 @@ PyObject *ActivePyModules::get_python(const std::string &what)
   }
 }
 
-int ActivePyModules::start_one(PyModuleRef py_module)
+void ActivePyModules::start_one(PyModuleRef py_module)
 {
   std::lock_guard l(lock);
 
   ceph_assert(modules.count(py_module->get_name()) == 0);
 
-  modules[py_module->get_name()].reset(new ActivePyModule(py_module, clog));
-  auto active_module = modules.at(py_module->get_name()).get();
-
-  int r = active_module->load(this);
-  if (r != 0) {
-    // the class instance wasn't created... remove it from the set of activated
-    // modules so commands and notifications aren't delivered.
-    modules.erase(py_module->get_name());
-    return r;
-  } else {
-    dout(4) << "Starting thread for " << py_module->get_name() << dendl;
-    active_module->thread.create(active_module->get_thread_name());
-
-    return 0;
-  }
+  const auto name = py_module->get_name();
+  modules[name].reset(new ActivePyModule(py_module, clog));
+  auto active_module = modules.at(name).get();
+
+  // Send all python calls down a Finisher to avoid blocking
+  // C++ code, and avoid any potential lock cycles.
+  finisher.queue(new FunctionContext([this, active_module, name](int) {
```

Comment:
what exactly is the cycle that is being avoided? it seems like before the finisher approach all of the calls were synchronous and any deadlock would have always been triggered.